akka-http/src/main/scala/akka/http/scaladsl/server/directives/ReverseProxyDirectives.scala

```diff
@@ -0,0 +1,142 @@
+/*
+ * Copyright (C) 2009-2016 Lightbend Inc. <http://www.lightbend.com>
+ */
+
+package akka.http.scaladsl.server.directives
+
+import akka.actor.ActorSystem
+import akka.http.scaladsl.Http
+import akka.http.scaladsl.Http.{ HostConnectionPool, OutgoingConnection }
+import akka.http.scaladsl.model.Uri.Authority
+import akka.http.scaladsl.model.{ HttpResponse, _ }
+import akka.http.scaladsl.model.headers._
+import akka.http.scaladsl.server._
+import akka.http.scaladsl.server.directives.BasicDirectives._
+import akka.http.scaladsl.server.directives.FutureDirectives._
+import akka.http.scaladsl.server.directives.RouteDirectives._
+import akka.pattern.CircuitBreaker
+import akka.stream._
+import akka.stream.scaladsl.{ Flow, Keep, Sink, Source }
+
+import scala.collection.immutable.Seq
+import scala.concurrent.Future
+import scala.util.{ Failure, Success, Try }
+
+trait ReverseProxyDirectives {
+  type RequestInterceptor = (HttpRequest) ⇒ HttpRequest
+  type ResponseInterceptor = ((HttpResponse) ⇒ HttpResponse)
+  type RequestExecutor = ((HttpRequest) ⇒ Future[HttpResponse])
+
+  def completeUnmatchedPathWithReverseProxy(authority: Authority, circuitBreakerOption: Option[CircuitBreaker] = None, responseInterceptor: ResponseInterceptor = identity[HttpResponse](_))(implicit system: ActorSystem): Route = {
+    forward(authority, circuitBreakerOption).tapply(x ⇒ complete(x._1.map(responseInterceptor)))
+  }
+
+  def forward(authority: Authority, circuitBreakerOption: Option[CircuitBreaker] = None)(implicit system: ActorSystem): Directive1[Try[HttpResponse]] = {
+    extractUri.flatMap { uri ⇒
+      val forwardUri: Uri = uri.withAuthority(authority)
+      forwardToUri(forwardUri, circuitBreakerOption)
+    }
+  }
+
+  def forwardUnmatchedPath(authority: Authority, circuitBreakerOption: Option[CircuitBreaker] = None)(implicit system: ActorSystem): Directive1[Try[HttpResponse]] = {
+    extractUnmatchedPath.flatMap { unmatched ⇒
+      extractUri.flatMap { uri ⇒
+        val forwardUri: Uri = uri.withAuthority(authority).withPath(unmatched)
+        forwardToUri(forwardUri, circuitBreakerOption)
+      }
+    }
+  }
+
+  def forwardToUri(uri: Uri, circuitBreakerOption: Option[CircuitBreaker] = None)(implicit system: ActorSystem): Directive1[Try[HttpResponse]] = {
```

Comment:
If you rebase your branch now you'll notice a new directive: https://github.com/akka/akka/pull/21097 `extractActorSystem`.
So you can pull out the system out of a hat here, instead of passing it through 👍 

I'll think about the other options now